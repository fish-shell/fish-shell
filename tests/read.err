
####################
# Read with no vars is an error
read: Expected at least 1 args, got only 0

####################
# Read with -a and anything other than exactly on var name is an error
read: Expected 1 args, got 0
read: Expected 1 args, got 2

####################
# Test splitting input

####################
# Test splitting input with IFS empty

####################
# read -n tests

####################
# read -z tests

####################
# Chunked read tests

####################
# Confirm reading non-interactively works -- #4206 regression

####################
# Test --delimiter (and $IFS, for now)

####################
# Delimiter =

####################
# Delimiter empty

####################
# IFS empty string

####################
# IFS unset

####################
# Delimiter =
