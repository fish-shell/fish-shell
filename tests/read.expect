# vim: set filetype=expect:

proc expect_read_prompt {} {
    expect -re "\\r\\n?read> $"
}

proc expect_marker {text} {
    expect_prompt -re "\\r\\n@MARKER:$text@\\r\\n" {} unmatched {
        abort "Couldn't find marker line '$text'"
    }
}

spawn $fish

expect_prompt

# read

send_line "read foo"
expect_read_prompt
send_line "text"
expect_prompt
print_var_contents foo

send_line "read foo"
expect_read_prompt
send_line "again\r_marker 1"
expect_prompt
expect_marker 1
print_var_contents foo

send_line "read foo"
expect_read_prompt
send_line -h "bar\r_marker 2"
expect_prompt
expect_marker 2
print_var_contents foo

# read -n

send_line "read -n 3 foo"
expect_read_prompt
send_line -h "123_marker 3"
expect_prompt
expect_marker 3
print_var_contents foo

send_line "read -n 3 foo"
expect_read_prompt
send_line "456_marker 4"
expect_prompt
expect_marker 4
print_var_contents foo

send_line "read -n 12 foo bar"
expect_read_prompt
send_line "hello world!_marker 5"
expect_prompt
expect_marker 5
print_var_contents foo
print_var_contents bar

send_line "bind ` 'commandline -i test'`"
expect_prompt
send_line "read -n 4 foo"
expect_read_prompt
send_line "te`_marker 6"
expect_prompt
expect_marker 6
print_var_contents foo

send_line "read -n 4 foo"
expect_read_prompt
send_line -h "12`_marker 7"
expect_prompt
expect_marker 7
print_var_contents foo
