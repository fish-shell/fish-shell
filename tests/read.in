#
# Test read builtin and IFS
#

count (echo one\ntwo)
set -l IFS \t
count (echo one\ntwo)
set -l IFS
count (echo one\ntwo)
echo [(echo -n one\ntwo)]
count (echo one\ntwo\n)
echo [(echo -n one\ntwo\n)]
count (echo one\ntwo\n\n)
echo [(echo -n one\ntwo\n\n)]
set -le IFS

function print_vars --no-scope-shadowing
	set -l space
	set -l IFS \n # ensure our command substitution works right
	for var in $argv
		echo -n $space (count $$var) \'$$var\'
		set space ''
	end
	echo
end

echo
echo 'hello there' | read -l one two
print_vars one two
echo 'hello there' | read -l one
print_vars one
echo '' | read -l one
print_vars one
echo '' | read -l one two
print_vars one two
echo 'test' | read -l one two three
print_vars one two three

echo
set -l IFS
echo 'hello' | read -l one
print_vars one
echo 'hello' | read -l one two
print_vars one two
echo 'hello' | read -l one two three
print_vars one two three
echo '' | read -l one
print_vars one
echo 't' | read -l one two
print_vars one two
echo 't' | read -l one two three
print_vars one two three
echo ' t' | read -l one two
print_vars one two
set -le IFS

echo
echo 'hello there' | read -la ary
print_vars ary
echo 'hello' | read -la ary
print_vars ary
echo 'this is a bunch of words' | read -la ary
print_vars ary
echo '   one   two     three' | read -la ary
print_vars ary
echo '' | read -la ary
print_vars ary

echo
set -l IFS
echo 'hello' | read -la ary
print_vars ary
echo 'h' | read -la ary
print_vars ary
echo '' | read -la ary
print_vars ary
set -le IFS
