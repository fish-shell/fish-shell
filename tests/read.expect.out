$foo: 'text'
$foo: 'again'
$foo: 'bar'
$foo: '123'
$foo: '456'
$foo: 'hello'
$bar: 'world!'
$foo: 'tete'
$foo: '12te'
